src/Microsoft.AspNetCore.Authorization/AuthorizationHandler.cs

```diff
@@ -6,38 +6,53 @@
 
 namespace Microsoft.AspNetCore.Authorization
 {
+    /// <summary>
+    /// Base class for authorization handlers that need to be called for a specific requirement type.
+    /// </summary>
+    /// <typeparam name="TRequirement">The type of the requirement to handle.</typeparam>
     public abstract class AuthorizationHandler<TRequirement> : IAuthorizationHandler
-        where TRequirement : IAuthorizationRequirement
+            where TRequirement : IAuthorizationRequirement
     {
-        public void Handle(AuthorizationHandlerContext context)
+        async Task IAuthorizationHandler.HandleAsync(AuthorizationHandlerContext context)
         {
             foreach (var req in context.Requirements.OfType<TRequirement>())
             {
-                Handle(context, req);
+                await HandleAsync(context, req);
             }
         }
 
-        public virtual async Task HandleAsync(AuthorizationHandlerContext context)
+        /// <summary>
+        /// Makes a decision if authorization is allowed based on a specific requirement.
+        /// </summary>
+        /// <param name="context">The authorization information.</param>
+        /// <param name="requirement">The requirement to evaluate.</param>
+        protected virtual void Handle(AuthorizationHandlerContext context, TRequirement requirement)
```

Comment:
Lets go with the more drastic but simpler for long term options.  Lets get rid of sync entirely and have only two methods, and rename the Handle to not have the same name.

Thoughts?
```
   public abstract AuthenticationHandler<TReq> : IAuthenticationHandler {
      public virtual Task HandleAsync(AuthenticationHandlerContext context);
      protected abstract Task HandleRequirementAsync(AuthenticationHandlerContext context, TReq requirement)
   }
```